code/modules/power/engines/singularity/field_generator.dm

```diff
@@ -177,49 +177,60 @@ GLOBAL_LIST_EMPTY(field_generator_fields)
 				start_fields()
 
 
-/obj/machinery/field/generator/proc/calc_power()
+/obj/machinery/field/generator/proc/calc_energy()
 	var/power_draw = 2 + length(fields)
 
 	if(draw_power(round(power_draw/2, 1)))
-		check_power_level()
+		spread_energy()
+		check_energy_level()
 		return 1
 	else
 		visible_message("<span class='danger'>[src] shuts down!</span>", "<span class='italics'>You hear something shutting down.</span>")
 		turn_off()
-		investigate_log("ran out of power and <font color='red'>deactivated</font>","singulo")
-		power = 0
-		check_power_level()
+		investigate_log("ran out of energy and <font color='red'>deactivated</font>","singulo")
+		energy = 0
+		check_energy_level()
 		return 0
 
-//This could likely be better, it tends to start loopin if you have a complex generator loop setup.  Still works well enough to run the engine fields will likely recode the field gens and fields sometime -Mport
-/obj/machinery/field/generator/proc/draw_power(draw = 0, failsafe = 0, obj/machinery/field/generator/G = null, obj/machinery/field/generator/last = null)
-	if((G && (G == src)) || (failsafe >= 8))//Loopin, set fail
+/**
+* Draws power. If there isn't enough energy to sustain the draw, draw from connected generators, up to 3 generators away.
+* We never do a 180, so at most we should be going around 270 degrees, and never loop.
+*
+* Arguments:
+** draw - Amount of energy needed to sustain powerdraw during this cycle
+** failsafe - Current depth of the recursion. We don't let this go above 3.
+** last - should be src of the previous call, we check against this to prevent going back and forth between two field generators.
+*/
+/obj/machinery/field/generator/proc/draw_power(draw = 0, failsafe = 0, obj/machinery/field/generator/last = null)
+	if(failsafe >= 3)// Asking at most 3 gens away so we can't loop.
 		return 0
 	else
 		failsafe++
 
-	if(power >= draw)//We have enough power
-		power -= draw
+	if(energy >= draw)// We have enough energy
+		energy -= draw
 		return 1
 
-	else//Need more power
-		draw -= power
-		power = 0
+	else//Need more energy
+		draw -= energy
+		energy = 0
 		for(var/CG in connected_gens)
 			var/obj/machinery/field/generator/FG = CG
-			if(FG == last)//We just asked you
+			if(FG == last)// We just asked you
 				continue
-			if(G)//Another gen is askin for power and we dont have it
-				if(FG.draw_power(draw,failsafe,G,src))//Can you take the load
-					return 1
-				else
-					return 0
-			else//We are askin another for power
-				if(FG.draw_power(draw,failsafe,src,src))
+			else// We are askin another for energy
+				if(FG.draw_power(draw, failsafe, src))
 					return 1
 				else
 					return 0
 
+/// Sends energy to every neighbour that has less energy
+/obj/machinery/field/generator/proc/spread_energy()
+	for(var/obj/machinery/field/generator/gen as anything in connected_gens)
+		if(energy > gen.energy)
```

Comment:
Might be best to make sure there is a certain amount of energy difference?

Reply:
I suppose it wouldn't hurt